### ISSUE TITLE
[Theme] Dark theme - Grayscale

### ISSUE THREAD

**Issue by gripped** [benbusby/whoogle-search#247]
While I'm a big fan of Whoogle, I'm not keen on the standard dark theme or what's been posted on the wiki.

An attempt at trying to avoid "bright light":

```    /* DARK THEME COLORS */
	--whoogle-dark-background: #18181a;
	--whoogle-dark-accent: #313131;
	--whoogle-dark-text: #bdbdbd;
	--whoogle-dark-contrast-text: #8e8e8e;
	--whoogle-dark-secondary-text: #8a8b8c;
	--whoogle-dark-result-bg: #313131;
	--whoogle-dark-result-title: #d9d9d9;
	--whoogle-dark-result-url: #eceff4;
	--whoogle-dark-result-visited: #959595;
```

What would help in improving this would be to separate whoogle-dark-accent / whoogle-accent into different elements for the logo and config background. i.e. I'd like a dark config background but then the logo becomes almost invisible.  In an ideal world I'd want the logo a tad brighter and the config bg much darker. Then I'd also go darker with whoogle-dark-result-bg, but I've matched  whoogle-dark-accent for consistency :)

Small point, and thanks for the great tool. I use it daily.

**Comment by benbusby** [comment]
Thanks for the kind words! There are a couple things that I wanted to clarify with you regarding your request.

For splitting out `whoogle-accent` from the config styling, are you meaning that you'd rather the "accent" color only be used for the logo itself? Or would a separate `whoogle-config` color be more fitting? Since the accent is also used for other UI components (like the search input & button) I just wanted to clarify what you're thinking. 

For example:
![image](https://user-images.githubusercontent.com/33362396/113158699-03710980-920a-11eb-99a2-4650258bea2e.png)

Let me know your thoughts and I'll try to work on a solution soon. 

> I can start a new issue for this if you prefer ?

I'll just include the fix in this issue, no need for a new one.

**Comment by gripped** [comment]
I think to me the optimal solution would be to separate all three of those. Maybe ditch `whoogle-accent` and have `whoogle-logo`, `whoogle-search-input` & `whoogle-config`. Or just the two `whoogle-logo` and one element , named whatever you think best for the config and search box combined.  `whoogle-accent` doesn't really say what it does. Not to me anyway.

IMHO the aim in any 'dark' theme is to have light text and dark backgrounds. The current arrangement allows that just about (except id="search-submit") but would be improved with more possibilities  for customization.

I don't think `<input type="submit" id="search-submit" value="Search">` should be using `--whoogle-dark-background`as the css for text. As that basically forces the text on that element to be dark. If it used `--whoogle-dark-contrast-text` they'd be more consistency.  Like this
![image](https://user-images.githubusercontent.com/3513053/113391765-efedac00-938b-11eb-99e0-b38dc523499d.png)
Then with separated `whoogle-logo` element I would brighten that a bit and make the search input (button and outline ) and config background (both currently `--whoogle-dark-accent`) darker.
Then it would be perfect !
I hope that makes some sense, even if you disagree.
As it is it's a vast improvement on the early days when dark was just a negative of light. But for a true dark theme dark text on brighter backgrounds must be banished :)

Thinking outside the box a little maybe having two sets of css, one light, one dark is overcomplicating things ?
Could just one set but with more separation be simpler? So no separate 'dark' theme css.  Because if I copy my dark colours into the light mode css my dark and light mode is identical, as you would expect, so the two lots of css is a bit superfluous. You could still have a dark mode selector. Which would pick from hardcoded css as appropriate, overridden by any Custom CSS:

**Comment by gripped** [comment]
Actually I take the last comment back. There are differences. Dropdown autocomplete is different at the very least. So there's more going on under the hood and maybe the two sets of css are necessary ?

**Comment by benbusby** [comment]
I had initially tried out only having one single CSS file, but the initial result was a bit too complicated as there are certain cases where additional dark theme styling is needed. I'm by no means a CSS expert though, so I'm sure there's room for general improvement. 

That all makes sense to me though! Ultimately I'm not very opinionated when it comes to theming, but a higher degree of customization seems like a good idea.

**Comment by benbusby** [comment]
I just pushed some changes that I think cover everything discussed, would you like to test it out before I merge? The changes are on branch `custom-css-tweaks`. Essentially `whoogle-accent` was split out into `whoogle-logo` and `whoogle-element-bg` (config / search field background), and `whoogle-background` was updated to `whoogle-page-bg` to distinguish itself a bit better from `whoogle-element-bg`. 

New default theme for reference:

```css
:root {
    /* LIGHT THEME COLORS */
    --whoogle-logo: #685e79;
    --whoogle-page-bg: #ffffff;
    --whoogle-element-bg: #685e79;
    --whoogle-text: #000000;
    --whoogle-contrast-text: #ffffff;
    --whoogle-secondary-text: #70757a;
    --whoogle-result-bg: #ffffff;
    --whoogle-result-title: #1967d2;
    --whoogle-result-url: #0d652d;
    --whoogle-result-visited: #4b11a8;

    /* DARK THEME COLORS */
    --whoogle-dark-logo: #685e79;
    --whoogle-dark-page-bg: #222222;
    --whoogle-dark-element-bg: #685e79;
    --whoogle-dark-text: #ffffff;
    --whoogle-dark-contrast-text: #000000;
    --whoogle-dark-secondary-text: #bbbbbb;
    --whoogle-dark-result-bg: #000000;
    --whoogle-dark-result-title: #1967d2;
    --whoogle-dark-result-url: #4b11a8;
    --whoogle-dark-result-visited: #bbbbff;
}
```

**Comment by gripped** [comment]
I am playing around with it. I'll get back to you once finished.

**Comment by gripped** [comment]
Line 80 of app/static/css/dark-theme.css sets --whoogle-contrast-text instead of --whoogle-dark-contrast-text.
I can create a PR if you wish but it's a tiny change.

In app/templates/header.html something is going wrong with the logic to check for dark mode I think ?
It's using whoogle-logo when dark mode is set instead of whoogle-dark-logo.
![whoogle2](https://user-images.githubusercontent.com/3513053/113479804-00d01780-9489-11eb-8b37-2d36ba3265ef.png)
I tried it in chrome as well, with zero extensions just in case, but same result
![whoogle4](https://user-images.githubusercontent.com/3513053/113479972-d6328e80-9489-11eb-88da-e1aea50c8ce6.png)


Had a quick look but have not figured it out yet (if ever)

**Comment by benbusby** [comment]
Thanks for pointing out those issues! Those should both be fixed now.

**Comment by gripped** [comment]
That's almost perfect for the home screen:

'<button id="config-collapsible" class="collapsible">Configuration</button>' uses `--whoogle-dark-element-bg`when collapsed, `--whoogle-dark-contrast-text` when active.
I think it should be `--whoogle-dark-contrast-text` in both cases. Otherwise, they way I intend to style it, the button is almost invisible when collapsed. Not a biggie though. I know it's there.

All the results pages 'All, News, etc' are using `--whoogle-logo` instead of `--whoogle-dark-logo`

**Comment by benbusby** [comment]
Hmmm. I see where you're coming from, but for other cases where the element and page background colors are higher contrast (say `#222222` for page background and `#bbbbbb` for element background, with `#ffffff` regular text and `#000000` for contrast text), it would make sense to have the contrast text be on the darker side, and only applied when the config menu is expanded. Otherwise the same result would happen in your case where the config expanding button is almost invisible. I don't really have a better solution at the moment, but I think that most themes would benefit from using the element background color for the config button text. Ultimately it would be beneficial to see how everyone else uses the the custom theming feature, and if there needs to be a separate field for "config menu button" or just adopting your proposed solution, I'm happy to do that instead.

> All the results pages 'All, News, etc' are using --whoogle-logo instead of --whoogle-dark-logo

Thanks for catching that, that should be fixed in the latest commit.

**Comment by gripped** [comment]
Personally I'll be patching my own Whoogle with the PR above.
This is my dark theme.
```
body {
    max-width: 85% !important;
}

:root {
    --whoogle-dark-logo: #888;
    --whoogle-dark-page-bg: #080808;
    --whoogle-dark-element-bg: #111;
    --whoogle-dark-text: #ddd;
    --whoogle-dark-contrast-text: #aaa;
    --whoogle-dark-secondary-text: #8a8b8c;
    --whoogle-dark-result-bg: #111;
    --whoogle-dark-result-title: #ddd;
    --whoogle-dark-result-url: #eceff4;
    --whoogle-dark-result-visited: #959595;
}

input {
  border-color: #080808;
  border-radius: 2px;
}

select {
  border-color: #080808;
  border-radius: 2px;
}

textarea {
  border-color: #080808;
  border-radius: 2px;
}
```
In conjunction with the PR above it looks like this:
![image](https://user-images.githubusercontent.com/3513053/113858332-81e62200-979b-11eb-88bd-911ef87dc799.png)
![image](https://user-images.githubusercontent.com/3513053/113858403-9aeed300-979b-11eb-855c-b4cc30e7edc0.png)
![image](https://user-images.githubusercontent.com/3513053/113858505-b954ce80-979b-11eb-971f-71fc40581b18.png)


Footer
![image](https://user-images.githubusercontent.com/3513053/113858597-cffb2580-979b-11eb-91b5-d4edc379be32.png)


At some point I intend to figure out how to how to restyle the checkboxes, text area scrollbar and select box buttons.

You mentioned before about not being opinionated about themes. I'm not. It's my eyes that are!
Until you added user customisable css Whoogle was being styled, for me, by the "Quantum nox - Global dark theme" Stylus theme. Now I can turn that off for whoogle as the provided dark theme produces a better result. So thanks again.

**Comment by gripped** [comment]
The select box buttons look far better with webkit over Firefox (above). So partly just a browser issue.
Chrome
![image](https://user-images.githubusercontent.com/3513053/113860597-1c476500-979e-11eb-804a-5e1f99e2d370.png)

**Comment by benbusby** [comment]
Oh nice. I really like how that looks. I'm curious, would it be a good enough compromise to just use `--whoogle(-dark)-text` for the config button text? I feel like that would make sense, since the regular (non `-contrast`) text is meant to be visible against the regular page background color anyways. Alternatively, the theme could just set the color for the button explicitly in the custom css, i.e.:

```css
#config-collapsible {
  color: var(--whoogle-dark-contrast-text) !important;
}
```

or even

```css
#config-collapsible:not(.active) {
  color: var(--whoogle-dark-contrast-text) !important;
}
```

Either way I'd personally like to use this theme, so I'd like to get your changes merged in with whatever resolution you prefer.

**Comment by gripped** [comment]
My rational was to match the text colour of `#config-collapsible` with the labels in `#config-div`.
But either works. Whatever you prefer.
As you point out it can be overridden in the custom css.

What I've been playing with the last hour or so trying to remove all hardcoded colours from dark-theme.css. It's mainly related to autocomplete. While it works it would be better (IMHO) if it used the css variables available. Especially 
```
 .autocomplete-items {
    border: 1px solid #685e79;
}
```
&
```
li a {
    color: #4b8eaa !important;
}
```
In my experiments so far I think it looks better, more consistent ?
My first attempt: (Still a bit away from how I want it in the end)
![image](https://user-images.githubusercontent.com/3513053/113898879-2b410e00-97c4-11eb-89ca-8581ced532d0.png)
Standard with hardcoding:
![image](https://user-images.githubusercontent.com/3513053/113899581-db167b80-97c4-11eb-9fe6-65391a870083.png)
Notice the hardcoded purple border which has no real place in this theme any more.
It will take awhile of playing with it before I'm happy but I think as an aim removing hardcoded colours from both the light and dark theme css is the way to go?

But I'd consider that a separate issue.
I'll keep playing around. I'll put my sunglasses on and try to make the light theme css follow the same format as the dark css I come up with.
When I'm happy with it (could be never!) I'll send another PR on a new branch.

I'm as far from an expert with css as it's possible to be but there's only one way to learn :)

The only other thing I'd say is if you do merge this branch it may have implications for those who have set custom css already and find that it no longer works, as some of the names have changed. So I guess you'd need to communicate that in some way? Your probably well aware of this but I thought I'd mention it.

**Comment by trumad** [comment]
I really love how this looks, and when you're finished I think it should be one of the default themes.

**Comment by benbusby** [comment]
@trumad Yes, my plan is to make this theme the default dark theme once everything is merged in. 

@gripped I merged your changes into the branch, and pushed a small change regarding our other discussion about which color to use for the unopened config button text (I opted for `--whoogle(-dark)-text` since it should provide enough contrast in any theme, regardless of light/dark preference, and I can add a note to the wiki for how to override that value if desired). Here is your theme with that change implemented. When the config menu is expanded, the button label matches the other labels in `#config-div`, but is slightly brighter when collapsed.

![image](https://user-images.githubusercontent.com/33362396/114049372-65a8ba80-9859-11eb-8135-45d858bfa5d9.png)
![image](https://user-images.githubusercontent.com/33362396/114049404-6ccfc880-9859-11eb-84a4-2b1ffad5984b.png)

With those changes, the new default color scheme for Whoogle would be:

```css
:root {
    /* LIGHT THEME COLORS */
    --whoogle-logo: #685e79;
    --whoogle-page-bg: #ffffff;
    --whoogle-element-bg: #685e79;
    --whoogle-text: #000000;
    --whoogle-contrast-text: #ffffff;
    --whoogle-secondary-text: #70757a;
    --whoogle-result-bg: #ffffff;
    --whoogle-result-title: #1967d2;
    --whoogle-result-url: #0d652d;
    --whoogle-result-visited: #4b11a8;

    /* DARK THEME COLORS */
    --whoogle-dark-logo: #888888;
    --whoogle-dark-page-bg: #080808;
    --whoogle-dark-element-bg: #111111;
    --whoogle-dark-text: #dddddd;
    --whoogle-dark-contrast-text: #aaaaaa;
    --whoogle-dark-secondary-text: #8a8b8c;
    --whoogle-dark-result-bg: #111111;
    --whoogle-dark-result-title: #dddddd;
    --whoogle-dark-result-url: #eceff4;
    --whoogle-dark-result-visited: #959595;
}
```

If this all looks good to you, I'll merge these changes in and update the wiki. I think overall improvements to the theming options and removing the remaining hardcoded colors can be done in separate issues -- I mostly just want to get the core set of variables finalized and hopefully have a new default dark theme that everyone likes better than my original (boring) one :)

**Comment by gripped** [comment]
It's looking very good.
And yes I agree about the hardcoding.

Thanks for being so reactive.

**Comment by glitsj16** [comment]
> At some point I intend to figure out how to how to restyle the checkboxes, text area scrollbar and select box buttons.

> The select box buttons look far better with webkit over Firefox (above). So partly just a browser issue.

IMO it's much nicer to leave styling of form elements up to the browser. When I comment out the background property on `textarea`, `input` and `select` selectors - both in dark-theme.css and light-theme.css - these form elements look _fully_ integrated and get whatever styling is used natively. Especially on Firefox this is a big improvement.

Did I mention this is a truly awesome project?! Thanks to all involved!

**Comment by benbusby** [comment]
Sorry, I guess I forgot to update this issue. @gripped your theme was merged and made the default "dark" theme as of 1030118d0b1c7786894ac57990018f04f83eee61, with hardcoded values removed just now in b7e48a9597bbcba39b228abd430d2b3df38259d8. If you encounter anything else (or come up with a light theme counterpart to your dark theme) of course feel free to open another issue! Thanks again for all of your help.

@glitsj16 I agree about the browser default styling of form elements, but at least for me (also using Firefox), it's still using a white background with black text for dark themes or "low contrast" light themes where that might not be appropriate. If there's some workaround that could be applied for all browsers I'd be happy to include that though. Or alternatively, trying to mimic the browser default form element styling with css.

> Did I mention this is a truly awesome project?! Thanks to all involved!

Thank you!

**Comment by glitsj16** [comment]
> @glitsj16 I agree about the browser default styling of form elements, but at least for me (also using Firefox), it's still using a white background with black text for dark themes or "low contrast" light themes where that might not be appropriate. If there's some workaround that could be applied for all browsers I'd be happy to include that though. Or alternatively, trying to mimic the browser default form element styling with css.

@benbusby Firefox is notoriously picky with regards to dark themes. I think the reason why you are seeing white background and black text for dark/low-contrast themes is your default GTK theme (probably Adwaita if you haven't change your defaults). FYI, there's a few things one can do on the `user` side to get Firefox to respect dark themes:
- start Firefox with a GTK theme of choice for dark form elements via environment var
- change settings in about:config

More info on these topics can be found on the [Firefox](https://wiki.archlinux.org/index.php/Firefox#Dark_themes) and [GTK](https://wiki.archlinux.org/index.php/GTK#Themes) Arch Linux wiki pages. Just referencing this here so users can find that info. If whoogle-search wants to ensure a uniform theming experience across all browsers/platforms, mimicing form elements is indeed the only option. The new dark theme looks great BTW.

**Comment by gripped** [comment]
@benbusby 
Yes I'd figured out the changes have been merged. Thanks. Have had a little play with removing the hardcoding but not happy yet.

@glitsj16 
I tried what you suggested. But it didn't match my system theme (KDE / Plasma, Breeze dark). It just resulted in white boxes with black text. It would be great if it did work like  that but throw in multiple operating systems and multiple browsers and I don't think it does ?
